internal/store/db/redis/store.go

```diff
@@ -16,37 +16,208 @@
 package redis
 
 import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
 	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/contracts"
 	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db"
 	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db/interfaces"
+	"github.com/edgexfoundry/app-functions-sdk-go/internal/store/db/redis/models"
+
+	"github.com/gomodule/redigo/redis"
 )
 
+var currClient *Client // a singleton so Readings can be de-referenced
+var once sync.Once
+
+const redisCollection = "store"
+
+// Client provides an implementation for the Client interface for Redis
+type Client struct {
+	Pool      *redis.Pool // A thread-safe pool of connections to Redis
+	BatchSize int
+}
+
 // Store persists a stored object to the data store.
-func Store(o contracts.StoredObject) (id string, err error) {
-	return "", nil
+func (c Client) Store(o contracts.StoredObject) (string, error) {
+	err := o.ValidateContract(false)
+	if err != nil {
+		return "", err
+	}
+
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	return addToDB(conn, o)
 }
 
 // RetrieveFromStore gets an object from the data store.
-func RetrieveFromStore(appServiceKey string) (objects []contracts.StoredObject, err error) {
-	return nil, nil
+func (c Client) RetrieveFromStore(appServiceKey string) (objects []contracts.StoredObject, err error) {
+	// do not satisfy requests for a blank ASK
+	if appServiceKey == "" {
+		return nil, errors.New("no AppServiceKey provided")
+	}
+
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	ids, err := redis.Values(conn.Do("SMEMBERS", appServiceKey))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ids) == 0 {
+		return nil, nil
+	}
+
+	values, err := redis.ByteSlices(conn.Do("MGET", ids...))
+	if err != nil {
+		return nil, err
+	}
+
+	var model models.StoredObject
+
+	for _, bytes := range values {
+		err = model.UnmarshalJSON(bytes)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, model.ToContract())
+	}
+
+	return objects, nil
 }
 
 // Update replaces the data currently in the store with the provided data.
-func Update(o contracts.StoredObject) error {
+func (c Client) Update(o contracts.StoredObject) error {
+	err := o.ValidateContract(true)
+	if err != nil {
+		return err
+	}
+
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	err = removeFromDB(conn, o)
+	if err != nil {
+		return err
+	}
+
```

Comment:
This feels quite heavy weight to me. The StoredObject will overwrite the pervious StoredObject when you use the set command. If the AppServiceKey has changed, it needs to be removed from the SET and the new AppServiceKey needs to be added.

Reply:
You're right. I'll update Update to unmarshal and take a look at the object.